File: binarytreepostordertraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> res;
        stack<TreeNode*> s;
        TreeNode* lastNode = NULL;
        while (root or !s.empty()) {
            if (root) {
                s.push(root);
                root = root -> left;
            }
            else {
                TreeNode* topNode = s.top();
                if (topNode -> right && lastNode != topNode -> right)
                    root = topNode -> right;
                else {
                    res.push_back(topNode -> val);
                    lastNode = topNode;
                    s.pop();
                }
            }
        }
        return res;  
    }
};

File: python/binarytreepostordertraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def postorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = []
        if not root:
            return res
        stack = []
        while root or stack:
            if root:
                res.insert(0, root.val)
                stack.append(root)
                root = root.right
            else:
                t = stack.pop()
                root = t.left
        return res
                
